torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp

```diff
@@ -122,6 +122,115 @@ class DomainMap : public pointwise_utils::DomainMap {
+// consumed by the satisfied one. We will split that dim and large dim and and
+// use the splitted ones to satisfy both of them
+void maybeBuildVirtualInnerDims(
+    TransposeParams& params,
+    const std::vector<int64_t>& shape,
+    int64_t inner_most1,
+    int64_t inner_most2) {
+  int64_t merged_size1 = shape[inner_most1];
+  int64_t merged_size2 = shape[inner_most2];
+
+  // merge inner_most1 and inner_most2 left until we are done or we can no
+  // longer do so
+  int64_t dim = inner_most1 - 1;
+  while (dim >= 0 && dim != inner_most2 && merged_size1 < params.tile_size1) {
+    params.dims_merged_with_1.push_back(dim);
+    merged_size1 *= shape[dim];
+    dim--;
+  }
+  dim = inner_most2 - 1;
+  while (dim >= 0 && dim != inner_most1 && merged_size2 < params.tile_size2) {
+    params.dims_merged_with_2.push_back(dim);
+    merged_size2 *= shape[dim];
+    dim--;
+  }
+  // If any of them are unsatisfied, then find other dims to merge
+  std::unordered_set<int64_t> unavailable_dims;
+  unavailable_dims.reserve(
+      2 + params.dims_merged_with_1.size() + params.dims_merged_with_2.size());
+  unavailable_dims.insert(inner_most1);
+  unavailable_dims.insert(inner_most2);
+  for (auto i : params.dims_merged_with_1) {
+    unavailable_dims.insert(i);
+  }
+  for (auto i : params.dims_merged_with_2) {
+    unavailable_dims.insert(i);
+  }
+  dim = shape.size() - 1;
+  while (dim >= 0 && merged_size1 < params.tile_size1) {
+    if (unavailable_dims.count(dim) == 0) {
+      params.dims_merged_with_1.push_back(dim);
+      merged_size1 *= shape[dim];
+      unavailable_dims.insert(dim);
+    }
+    dim--;
+  }
+  dim = shape.size() - 1;
+  while (dim >= 0 && merged_size2 < params.tile_size2) {
+    if (unavailable_dims.count(dim) == 0) {
+      params.dims_merged_with_2.push_back(dim);
+      merged_size2 *= shape[dim];
+      unavailable_dims.insert(dim);
+    }
+    dim--;
+  }
+  // If both are satisfied, then we are done. If neither are satisfied, then it
+  // is impossible to satisfy both of them, also done.
+  if ((merged_size1 < params.tile_size1) ==
+      (merged_size2 < params.tile_size2)) {
+    return; // no need to split
+  }
+  // If one of them are not satisfied, then split the last merged dim in the
+  // other one
+  int64_t large_dim;
+  int64_t split_factor;
+  if (merged_size1 < params.tile_size1) {
+    large_dim = params.dims_merged_with_2.back();
+    split_factor = ceilDiv(params.tile_size1, merged_size1);
+  } else {
+    large_dim = params.dims_merged_with_1.back();
+    split_factor = ceilDiv(params.tile_size2, merged_size2);
+  }
+  params.split_before_tiling.push_back({large_dim, split_factor});
+  // adjust all dims to after-split
+  for (auto& i : params.dims_merged_with_1) {
+    if (i > large_dim) {
+      i++;
+    }
+  }
+  for (auto& i : params.dims_merged_with_2) {
+    if (i > large_dim) {
+      i++;
+    }
+  }
+  // give the split-out dim to the unsatisfied one
```

Comment:
Okay, I think I finally understand. You're trying to split out a portion of the larger dimension to add to the smaller dimension. So if group1's inner dimension is smaller than the selected tile, you'll attempt to pull dimensions into it. If you can't find other dimensions to put into group 1, you'll take dimensions from group 2. You started "maybe merging" dimensions from the inner most to the outer most dimension. With group 1 dimensions being inside group 2 dimensions. Looks like the ordering of the dimensions is based on group1 (reference 1). So if you take a portion of group 2 for group 1 it will be the outer most dimension.

Reply:
For this approach why wouldn't you: Take in dimensions into group 1 based on group 1 ordering until you hit the tile size. Then if you have anything left and group 2 needs dimensions you take more dimensions into group 2 based on group 2 ordering.

For example:
Group 1 [i0{4}, i1{3}, i2{2}, i3{16}]
Group 2 [i1{3}, i0{4}, i3{16}, i2{2}]
With a tile of 32x32 you'd group 1 as: I1{3}, i3{16}
and group 2 as i0{4}, i2{2}

Reply:
Maybe the concern is group 1 and group 2 might not even be consistently ordered, which I think would be fair. I also think we don't know exactly how often this complicated of a case will even present itself so it's probably good enough as is.

Reply:
It is true that group 1 and group 2 might not be consistently ordered, but I still think the algorithm you proposed would generally provide better performance than the one I implemented. The only reason that I implemented it this way is mainly for simplicity. For example, I don't need to worry about the case where not all IDs are mapped from ref1 to ref2 (such as when a view exist). I would suggest leaving this as is and improving only if we see some real application that needs this.

Reply:
Agreed.